Add storage for TestSuitesSource

nulls opened the pull request:
- storage for test suites source
- combined standard and git test suites

nulls commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/testsuite/TestSuiteType.kt:
```diff
@@ -7,11 +7,6 @@ import kotlinx.serialization.Serializable
  */
 @Serializable
 enum class TestSuiteType {
```
can be removed

nulls replied:
removed

nulls commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/testsuite/TestSuitesSourceType.kt:
```diff
@@ -0,0 +1,10 @@
+ */
+enum class TestSuitesSourceType {
+    GIT,
+    STANDARD,
```
can be avoided

nulls replied:
removed

nulls commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/utils/GitUtil.kt:
```diff
@@ -177,3 +208,58 @@ private fun checkout(git: Git, branchOrCommit: String, setCreateBranchFlag: Bool
+
+private fun credentialsProvider(username: String?, password: String?): CredentialsProvider =
+    if (username != null && password != null) {
+        listOf<Any>().single()
```
debug code

nulls replied:
cleaned

nulls commented on db/v-2/tables/test_suite.xml:
```diff
@@ -15,4 +15,26 @@
     </changeSet>
 
+    <changeSet id="add-test-suites-source" author="nulls">
+        <delete tableName="test_execution" />
```
agreed with @akuleshov7 to clean old data in database

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesController.kt:
```diff
@@ -95,15 +95,6 @@ class TestSuitesController(
-     * @param testSuiteDtos suites, which need to be marked as obsolete
-     * @return response entity
-     */
-    @PostMapping("/internal/markObsoleteTestSuites")
```
now, not need to mark test suites as old, but we need to able to remove test suites by version

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/OrganizationController.kt:
```diff
@@ -184,6 +201,13 @@ internal class OrganizationController(
         .defaultIfEmpty(ResponseEntity.status(HttpStatus.FORBIDDEN).build())
 
+    @GetMapping("/internal/organization/{name}")
+    fun getByNameAsMono(@PathVariable name: String): Mono<Organization> = organizationService.findByName(name)
```
removed

nulls replied:
removed

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/OrganizationController.kt:
```diff
@@ -99,17 +103,30 @@ internal class OrganizationController(
      * @param authentication an [Authentication] representing an authenticated request
      * @return response
      */
     @PostMapping("/{organizationName}/update")
```
`{organizationName}` is set in url, but wasn't used

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,149 @@
+        @PathVariable name: String
+    ): Mono<TestSuitesSourceDto> =
+            Mono.just(organizationName)
+                .flatMap { organizationService.findByName(it).toMono() }
```
getByName for organization, it should exist here already

nulls replied:
done

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,149 @@
+        @RequestParam gitUrl: String,
+        @RequestParam testRootPath: String,
+        @RequestParam branch: String,
+    ): Mono<TestSuitesSource> = Mono.just(organizationName)
```
Can return DTO here

nulls replied:
changed

nulls commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/ExecutionRequest.kt:
```diff
@@ -58,8 +77,8 @@ data class ExecutionRequestForStandardSuites(
+    override val execCmd: String?,
+    override val batchSizeForAnalyzer: String?,
+    override val executionId: Long?,
     val version: String?,
```
version can be removed or reused

nulls replied:
removed, now it's set from test suite ids

nulls commented on save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/ArchiveUtils.kt:
```diff
@@ -0,0 +1,60 @@
+import kotlin.io.path.outputStream
+
+const val TAR_EXTENSION = ".tar"
+
```
Need to confirm tar archive with team

nulls commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/controllers/DownloadProjectController.kt:
```diff
@@ -126,29 +117,114 @@ class DownloadProjectController(
-                        testSuiteIds,
-                        execCmd,
-                        batchSizeForAnalyzer,
+            Flux.fromIterable(standardTestSuitesRepo.testRootPaths)
```
We don't update standard test suites on each run in current logic, need to spik fetching here

nulls replied:
added search of latest version

nulls commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/controllers/DownloadProjectController.kt:
```diff
@@ -126,29 +117,114 @@ class DownloadProjectController(
+        .flatMap { testSuites ->
+            updateExecution(
+                requestBase.project,
+                testSuites.map { it.version }
```
Current logic doesn't expect different version for test suites on a single execution run

nulls replied:
added validation

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,149 @@
+    fun getTestSuites(
+        @PathVariable organizationName: String,
+        @PathVariable name: String,
+        @PathVariable version: String,
```
done

nulls commented on save-backend/build.gradle.kts:
```diff
@@ -18,10 +18,6 @@ openApi {
     waitTimeInSeconds.set(120)
-
-    tasks.named<BootRun>("bootRun") {
-        jvmArgs("-Dbackend.fileStorage.location=\${HOME}/cnb/files")
```
doesn't work on windows

nulls commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/ExecutionRequest.kt:
```diff
@@ -21,6 +21,14 @@ sealed class ExecutionRequestBase {
     abstract val sdk: Sdk
+
+    /**
+     * Execution id of request, it's null on first request
```
leave as is to avoid additional changes

nulls commented on save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/storage/AbstractFileBasedStorage.kt:
```diff
@@ -109,6 +109,7 @@ abstract class AbstractFileBasedStorage<K>(
 
     companion object {
         const val DEFAULT_PROJECT_LOCATION = "default"
+        private const val KEY_SUFFIX = ".key"
```
need to remove, will be done in another item

Cheshiriks commented on db/v-2/insert-test-data/db.changelog-insert.xml:
```diff
@@ -7,6 +7,7 @@
 
     <include file="user-insert.xml" relativeToChangelogFile="true"/>
     <include file="organization-insert.xml" relativeToChangelogFile="true"/>
+    <include file="test_suites_source-insert.xml" relativeToChangelogFile="true"/>
```
Let's use as a separator in the file name only _-_
_test-suites-source-insert.xml_

nulls replied:
idea was to use format `<table_name>-insert.xml` similar to liquibase change logs for tables: `test_suite.xml` and `test_execution.xml`.
But as I can see, we use both formats: `lnk-contest-execution.xml`, `lnk-contest-project.xml` and etc.

Do I need to rename only this file or also `test_suites_source.xml`?

Cheshiriks replied:
I don't think it's essential, but I would prefer using the hyphen "-" because it doesn't require additional keystrokes on the keyboard.
It's probably not worth renaming old files anymore, so as not to break the changeSet.

nulls replied:
since `-` is more preferable, named `test_suites_source.xml` and `test_suites_source-insert.xml`

nulls commented on save-orchestrator/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:
```diff
@@ -335,7 +334,8 @@ class AgentService(
             .bodyToMono<TestSuite>()
             .map { testSuite ->
-                testSuite.type == TestSuiteType.STANDARD
+                // FIXME: a hardcoded value of url for standard test suites
```
removed

nulls commented on save-orchestrator/src/main/kotlin/com/saveourtool/save/orchestrator/Utils.kt:
```diff
@@ -94,23 +94,27 @@ internal fun DockerClient.findImage(imageId: String, meterRegistry: MeterRegistr
-    val batchSizeForTomlConfig = if (batchSizeForAnalyzer.isNullOrBlank()) {
-        ""
+fun createSyntheticTomlConfig(execCmd: String?, batchSizeForAnalyzer: String?): String? {
+    if (execCmd.isNullOrBlank() && batchSizeForAnalyzer.isNullOrBlank()) {
```
creates for all cases to avoid loading test suites from different test suites sources (standard tests case)

nulls commented on save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/ArchiveUtils.kt:
```diff
@@ -0,0 +1,48 @@
+ */
+@Suppress("NestedBlockDepth")
+fun Path.compressAsZipTo(targetPath: Path) {
+    Archiver().create(ArchiveStreamFactory.ZIP, targetPath, this)
```
Expander is used here

nulls commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/StandardTestSuiteUtils.kt:
```diff
@@ -7,9 +7,15 @@ package com.saveourtool.save.utils
+/**
+ * Directory in docker image, where all requested test suites will be located for execution
+ */
+const val TEST_SUITE_DIR = "test-suites"
```
moved

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/service/TestSuitesSourceService.kt:
```diff
@@ -0,0 +1,134 @@
+            branch: String,
+            subDirectory: String
+        ): String = buildString {
+            append(url.replace("https?://".toRegex(), ""))
```
it works

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/storage/TestSuitesSourceSnapshotStorage.kt:
```diff
@@ -0,0 +1,111 @@
+package com.saveourtool.save.backend.storage
```
removed

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/storage/TestSuitesSourceSnapshotStorage.kt:
```diff
@@ -0,0 +1,90 @@
+    override fun buildKey(rootDir: Path, pathToContent: Path): TestSuitesSourceSnapshotKey {
+        val pathNames = pathToContent.pathNamesTill(rootDir)
+        return TestSuitesSourceSnapshotKey(
+            pathNames[3],
```
suppressed

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/TestSuitesSourceRepository.kt:
```diff
@@ -0,0 +1,39 @@
+     * @param git
+     * @return found entities
+     */
+    fun findAllByGit(git: Git): List<TestSuitesSource>
```
removed this method

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/GitRepository.kt:
```diff
@@ -21,4 +21,10 @@ interface GitRepository : BaseEntityRepository<Git> {
+     * @param url
+     * @return all [Git] entities by [Git.url]
+     */
+    fun findAllByUrl(url: String): List<Git>
```
removed this method

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesController.kt:
```diff
@@ -72,14 +54,36 @@ class TestSuitesController(
+     * @param ids list of test suite ID
+     * @return response with names of test suite with id from provided list
+     */
+    @PostMapping("/internal/test-suite/names-by-ids")
```
phase 3

akuleshov7 replied:
pls add it to the issue

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:
```diff
@@ -324,29 +332,25 @@ class ExecutionController(private val executionService: ExecutionService,
-        .distinct()
-        .single()
-
     private fun Execution.getTestSuiteRepoUrl(): String = parseAndGetTestSuiteIds()
```
it's used for rerun -- need to change to using test suite ids directly

nulls replied:
will be handled in phase 3

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:
```diff
@@ -263,12 +267,16 @@ class ExecutionController(private val executionService: ExecutionService,
     @GetMapping(path = ["/api/$v1/getTestRootPathByExecutionId"])
```
it's used for rerun and default values for custom test run. will be removed in phase 3

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:
```diff
@@ -138,7 +138,7 @@ internal class ContestController(
         val test = testService.findTestsByTestSuiteId(testSuite.requiredId()).firstOrNull()
             ?: return Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND))
-        return TestFilesRequest(test.toDto(), testSuite.testRootPath).let<TestFilesRequest, Mono<TestFilesContent>> {
+        return TestFilesRequest(test.toDto(), testSuite.source.testRootPath).let<TestFilesRequest, Mono<TestFilesContent>> {
```
it's part of logic of contest, will be refactored

nulls replied:
supported storage here

nulls commented on .run/save-backend not docker.run.xml:
```diff
@@ -5,6 +5,7 @@
     </envs>
     <option name="MAIN_CLASS_NAME" value="com.saveourtool.save.backend.SaveApplicationKt" />
     <module name="save-cloud.save-backend.main" />
+    <shortenClasspath name="NONE" />
```
done

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/GitRepository.kt:
```diff
@@ -21,4 +21,10 @@ interface GitRepository : BaseEntityRepository<Git> {
+     * @param url
+     * @return all [Git] entities by [Git.url]
+     */
+    fun findAllByUrl(url: String): List<Git>
```
removed

petertrr commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,262 @@
+        .flatMap {
+            organizationService.findByName(it).toMono()
+        }
+        .switchIfEmptyToResponseException(HttpStatus.CONFLICT) {
```
Why `conflict`?

petertrr replied:
`NOT_FOUND` can be indeed confused with the case when test suite source by requested parameters is not found. Ok, let's keep `CONFLICT` for missing organizations

kgevorkyan commented on save-backend/build.gradle.kts:
```diff
@@ -18,10 +17,6 @@ openApi {
     outputFileName.set("backend-api-docs.json")
     waitTimeInSeconds.set(120)
-
-    tasks.named<BootRun>("bootRun") {
```
It's was required basically for CI, which prohibits to use our default directory `/home/cnb/...`
Probably, you can add some flags for Windows, which will disable it on this OS, and not remove this at all

nulls replied:
will revert

petertrr commented on save-preprocessor/src/main/resources/TestSuitesRepos:
```diff
@@ -1 +0,0 @@
-https://github.com/saveourtool/save-cli origin/main examples/kotlin-diktat;examples/discovery-test
```
Why is this file deleted? I can see references to `TestSuiteRepos.json`, but no new file. What is a strategy to work with standard suites in phase 3? Or is it already described in the issue?

nulls replied:
standard test suites will be added to `CQFN.org` as public tests and will be available to select for all users
right now, I move all these constants to hardcoded values on backend side. can extract to a file, but it will be not applicable in phase 3.

Also, will describe steps for phase 3 in #250

kgevorkyan commented:
Also it will be nice, if you can provide more detailed description about changes for such big PR

petertrr commented on save-cloud-common/build.gradle.kts:
```diff
@@ -44,10 +44,12 @@ kotlin {
             dependencies {
                 implementation(project.dependencies.platform(libs.spring.boot.dependencies))
                 implementation(libs.spring.security.core)
+                implementation(libs.spring.web)
```
What does this dependency provide?
Edit: OK, found, it's for common methods to get HTTP statuses

petertrr commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/testsuite/TestSuiteType.kt:
```diff
@@ -1,25 +0,0 @@
-    /**
-     * Type Obsolete Standard for old test suites
-     */
-    OBSOLETE_STANDARD,
```
What will happen with the periodic task for test suites update?

nulls replied:
they will create a new version, and we will take that latest available test suites source for standard run.
clean-up of old version will be done in phase 3 (or 4 :smile:)

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/service/GitService.kt:
```diff
@@ -4,12 +4,15 @@ import com.saveourtool.save.backend.repository.GitRepository
  * Service of git
  */
 @Service
+@RestController
```
yeap, thanks. will remove

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/service/TestSuitesSourceService.kt:
```diff
@@ -0,0 +1,132 @@
+    @Transactional
+    fun getStandardTestSuitesSources(): List<TestSuitesSource> {
+        // FIXME: a hardcoded values for standard test suites
+        // Will be removed in phase 3
```
yeap, in phase 3: standard test suites will be a public tests provided by organization `CQFN.org`

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/storage/TestSuitesSourceSnapshotStorage.kt:
```diff
@@ -0,0 +1,150 @@
+        testSuitesSourceName: String,
+    ): Mono<String> = list()
+        .filter { it.equalsTo(organizationName, testSuitesSourceName) }
+        .reduce { max, next ->
+            if (max.creationTimeInMills > next.creationTimeInMills) max else next
+        }
```
will change.

jfyi: Initially it was done on Flux<T> and it failed with error about MathFlux is not available

nulls replied:
removed the method -- we don't use it

kgevorkyan commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/storage/DebugInfoStorage.kt:
```diff
@@ -42,18 +45,13 @@ class DebugInfoStorage(
-        val executionId = folderNames[4].toLong()
+        val pathNames = pathToContent.pathNamesTill(rootDir)
+
+        val testName = pathNames[0].dropLast(SUFFIX_FILE_NAME.length)
+        val testLocation = pathNames[1]
+        val testSuiteName = pathNames[2]
+        val pluginName = pathNames[3]
+        val executionId = pathNames[4].toLong()
```
May be some data class for it?

nulls replied:
will add `DebugInfoStorageKey`

nulls replied:
added

nulls commented:
# Overview
Introduced a new entity `TestSuitesSource`. It's a source to fetch and detect test suites and tests.
`TestSuitesSource` describes only where tests are placed
A snapshot of `TestSuitesSource` is a zip archive of files are pulled from source with some specific `version`.

So,
`TestSuitesSource` is git credentials (`Git` entity) + branch + relative path in repo.
`Version` is sha1 or hash commit in git repository.

We store snapshots on backend side (it saves them to file system), it will be moved to S3 in future.

The main goal of this task is to add storage for TestSutiesSource: we split fetching test suites with files and run them.
Currently, the UI is not changed -- user still can choice standard tests suites or provide git url for personal tests.

For seamless transition, we create `TestSuitesSource` on the fly. User inputs git-url, git-branch and testRootPath and we search source with such values or create a new one (the same approach for standard test suites, values are hardcoded)

# How run looks now:
## Standard test suites
They still fetch periodically by preprocessor. We detect test suites and save them in backend + save snapshot of repository

When user runs a new run on standard test suites, we take the latest available version on backend (max by creation time of commit) and run common run using this version

## Git test suites
User can input branch or commit\hash\sha1.
If user inputs branch, we search or create a new test source for it and then detect a latest commit in git repository. Then we fetch test suites for this version and then trigger a common run using this version
If user inputs version, we fetch test suites for this version and then trigger a common run using this version

## Common run
We put selected testSuiteIds on execution and execute tests for them (create TestExecution), then initialize agent.
Orchestrator downloads additional files (binary for evaluated tools) + test suites sources to a tmp folder. Then orchestrator creates volume using this folder.


### Orchestrator download logic
Orchestrator create a tmp directory in format `/save-execution-{id}`. Then create a sub folder '/save-execution-{id}/test-suites' and download additional files there. Orchestrator download all test suites source snapshot to folder with name in format `/save-execution-{id}/test-suites/test-suites-source-{index}` (index can be changed to test suites source id in future to simplify debugging). Then orchestrator creates a synthecit toml config in `/save-execution-{id}/test-suites/save.toml` and run save cli with parameter `--include-suites`

kgevorkyan commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/storage/DebugInfoStorage.kt:
```diff
@@ -27,34 +30,30 @@ class DebugInfoStorage(
-        val executionId = folderNames[4].toLong()
-        return Pair(
+    override fun buildKey(rootDir: Path, pathToContent: Path): DebugInfoStorageKey {
+        val pathNames = pathToContent.pathNamesTill(rootDir)
```
Generally I meant this conversation from `pathNames` to set of fields.
I also see, that we use similar logic in other places, i.e. for `TestSuitesSourceSnapshotKey`
May be we can extract this logic into some method at least?
And not write `pathNames[1], pathNames[2], pathNames[3]...` everywhere?

nulls replied:
this logic of `partNames` depends on storage implementation: it depends on how we put files in file system.
it can be moved to `*Key` implementation as one of constructor (which takes list of string).  but we just move this logic to another place.

As another idea: we can create `*.metadata` file together with uploaded file. It also allows to store additional info without adding it to path. It fits to S3 which has metadata for objects. But I would like to leave this to next phase (I will need to add commit message to snapshots of test suites source), since it will require prepare transition steps

For now, I can re-write it using destructing declaration:
```suggestion
        val (testName, testLocation, testSuiteName, pluginName, executionId) = pathNames
        return DebugInfoStorageKey(
            executionId.toLong(),
            TestResultLocation(testSuiteName, pluginName, testLocation, testName.dropLast(SUFFIX_FILE_NAME.length))
        )
```

kgevorkyan replied:
Yeah, destructing declaration looks good for now

kgevorkyan commented on save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/testsuite/TestSuitesSourceDto.kt:
```diff
@@ -0,0 +1,30 @@
+typealias TestSuitesSourceDtoList = List<TestSuitesSourceDto>
+
+/**
+ * @param organizationName
```
Shouldn't diktat trigger on it? We have both `@param` and `@property` for one field
Probably, related https://github.com/saveourtool/diktat/issues/1074

nulls replied:
actually it's my fault. I added `@param` manually, but it's `@property` -- diktat adds the proper annotation, but doesn't remove the wrong and even doesn't raise an issue

nulls replied:
added comment to https://github.com/saveourtool/diktat/issues/1074

sanyavertolet commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,262 @@
+ */
+@RestController
+@RequestMapping("/internal/test-suites-sources", "/api/$v1/test-suites-sources")
+class TestSuitesSourceController(
```
Could you please write all the comments in swagger style

nulls replied:
added

akuleshov7 commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/configs/WebConfig.kt:
```diff
@@ -32,6 +33,7 @@ class WebConfig {
     fun jackson2ObjectMapperBuilderCustomizer() = Jackson2ObjectMapperBuilderCustomizer { jacksonObjectMapperBuilder: Jackson2ObjectMapperBuilder ->
         jacksonObjectMapperBuilder
             .mixIn(TestStatus::class.java, TestStatusMixin::class.java)
+            .featuresToDisable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
```
why needed in Jackson?

nulls replied:
I passed creation time as LocalDateTime, but I changed to long. Will revert

akuleshov7 commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesController.kt:
```diff
@@ -72,14 +54,36 @@ class TestSuitesController(
+     * @return response with names of test suite with id from provided list
+     */
+    @PostMapping("/internal/test-suite/names-by-ids")
+    fun findAllTestSuiteNamesByIds(@RequestBody ids: List<Long>) = ids
```
why do we need to find multiple test suites by ids?

nulls replied:
it's needed for orchestrator to setup `--include-suites` on orchestrator side. there we have only Execution with testSuiteIds

nulls replied:
but we don't need the next method -- will remove

akuleshov7 commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:
```diff
@@ -0,0 +1,262 @@
+                .contentType(MediaType.APPLICATION_OCTET_STREAM)
+                .body(testSuitesSourceSnapshotStorage.download(it))
+        }
+        .onErrorReturn(
```
`switchIfEmptyToNotFound` is not applicable here?

nulls replied:
will add

nulls replied:
added for case when key is not found

akuleshov7 commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/service/GitPreprocessorService.kt:
```diff
@@ -0,0 +1,126 @@
+typealias ArchiveProcessor<T> = (Path) -> Mono<T>
+
+/**
+ * Additional service for Git based [com.saveourtool.save.entities.TestSuitesSource]s
```
Don't get - why do we need this Service? 
For example: `cloneAndProcessDirectory` - I thought we already had such logic in preprocessor. Or how we download repositories?

nulls replied:
I extracted it to a separate Service to has access to configProperties and create a tmp folder

akuleshov7 commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/service/GitPreprocessorService.kt:
```diff
@@ -0,0 +1,126 @@
+        repositoryProcessor: GitRepositoryProcessor<T>,
+    ): Mono<T> {
+        val cloneAction: () -> Pair<Path, Instant> = {
+            val tmpDir = createTempDirectoryForRepository()
```
is it possible to move this code inside of Mono?

nulls replied:
it's a callable -- initialization of Mono.  

I use here `Mono.usingWhen` to control a tmp file which should be removed after all operations on this tmp file (kdoc describes it)

akuleshov7 commented:
lgtm. No chance that we can properly review this PR, so probably we need to merge it

akuleshov7 commented:
> # Overview
> Introduced a new entity `TestSuitesSource`. It's a source to fetch and detect test suites and tests. `TestSuitesSource` describes only where tests are placed A snapshot of `TestSuitesSource` is a zip archive of files are pulled from source with some specific `version`.
> 
> So, `TestSuitesSource` is git credentials (`Git` entity) + branch + relative path in repo. `Version` is sha1 or hash commit in git repository.
> 
> We store snapshots on backend side (it saves them to file system), it will be moved to S3 in future.
> 
> The main goal of this task is to add storage for TestSutiesSource: we split fetching test suites with files and run them. Currently, the UI is not changed -- user still can choice standard tests suites or provide git url for personal tests.
> 
> For seamless transition, we create `TestSuitesSource` on the fly. User inputs git-url, git-branch and testRootPath and we search source with such values or create a new one (the same approach for standard test suites, values are hardcoded)
> 
> # How run looks now:
> ## Standard test suites
> They still fetch periodically by preprocessor. We detect test suites and save them in backend + save snapshot of repository
> 
> When user runs a new run on standard test suites, we take the latest available version on backend (max by creation time of commit) and run common run using this version
> 
> ## Git test suites
> User can input branch or commit\hash\sha1. If user inputs branch, we search or create a new test source for it and then detect a latest commit in git repository. Then we fetch test suites for this version and then trigger a common run using this version If user inputs version, we fetch test suites for this version and then trigger a common run using this version
> 
> ## Common run
> We put selected testSuiteIds on execution and execute tests for them (create TestExecution), then initialize agent. Orchestrator downloads additional files (binary for evaluated tools) + test suites sources to a tmp folder. Then orchestrator creates volume using this folder.
> 
> ### Orchestrator download logic
> Orchestrator create a tmp directory in format `/save-execution-{id}`. Then create a sub folder '/save-execution-{id}/test-suites' and download additional files there. Orchestrator download all test suites source snapshot to folder with name in format `/save-execution-{id}/test-suites/test-suites-source-{index}` (index can be changed to test suites source id in future to simplify debugging). Then orchestrator creates a synthecit toml config in `/save-execution-{id}/test-suites/save.toml` and run save cli with parameter `--include-suites`

This should be in some developer's readme 😄 
Or inside the code - for example as a Kdoc to some of your classes

kgevorkyan commented on save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/utils/GitUtil.kt:
```diff
@@ -5,175 +5,106 @@
-        UsernamePasswordCredentialsProvider(gitDto.username, gitDto.password)
-    } else {
-        CredentialsProvider.getDefault()
+fun GitDto.detectDefaultBranchName() = Git.lsRemoteRepository()
```
Okay, basically you just separate this method into two

>`Also, we clone a repo to tmp file, so we need a pull method.`

But I can't see `pull` method

nulls replied:
sorry, typo: `so we don't need a pull method`

kgevorkyan replied:
oh, ok

nulls commented on save-backend/src/main/kotlin/com/saveourtool/save/backend/service/TestSuitesService.kt:
```diff
@@ -77,51 +76,33 @@ class TestSuitesService(
      */
-    fun getStandardTestSuites() =
-            testSuiteRepository.findAllByTypeIs(TestSuiteType.STANDARD).map { it.toDto() }
+    fun getStandardTestSuites() = testSuitesSourceService.getStandardTestSuitesSources()
```
need to find only latest one -- done
